url.go

```diff
@@ -211,6 +226,61 @@ func parseReference(url string) (*Reference, error) {
 		}
 		break
 	}
+	if r.User != "" {
+		if !names.IsValidUser(r.User) {
+			return nil, fmt.Errorf("entity URL has invalid user name: %q", url)
+		}
+	}
+	if !IsValidName(r.Name) {
+		return nil, fmt.Errorf("entity URL has invalid entity name: %q", url)
+	}
+	return &r, nil
+}
+
+func parseV2Reference(url *gourl.URL) (*Reference, error) {
+	var r Reference
+	r.Schema = "cs"
+	parts := strings.Split(strings.Trim(url.Path, "/"), "/")
+	if len(parts) == 0 {
+		return nil, fmt.Errorf("entity URL malformed, no parts: %q", url)
+	}
+	if parts[0] == "u" {
+		if len(parts) < 3 {
+			return nil, fmt.Errorf("entity URL malformed, expecting user and name: %q", url)
+		}
+		r.User = parts[1]
```

Comment:
an idiom that can look nice:

    r.User, parts = parts[1], parts[2:]

(and below)
YMMV.